# Load Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# load and read Dataset

In [2]:
from google.colab import files
uploaded= files.upload()

Saving ITI_Statistics_CaseStudy_Data.xlsx to ITI_Statistics_CaseStudy_Data (11).xlsx


In [3]:
df_raw=pd.read_excel('ITI_Statistics_CaseStudy_Data.xlsx', sheet_name="raw_data")

#  ETL Process

In [4]:
df_raw.shape

(271, 15)

In [5]:
df_raw.head()

,student_id,branch,track,gender,age,study_hours_week,commute_minutes,sleep_hours,projects_submitted,satisfaction_1to5,has_laptop,transport,pets_at_home,favorite_sport,final_score
0,ITI-1031,Cairo,Data Science,Male,21,16.5,34.1,5.5,2,3,Yes,Car,0,Swimming,79.3
1,ITI-1117,Assiut,Data Science,Female,29,25.1 hours,5.0,6.6,3,4,Yes,Metro,1,Football,91.7
2,ITI-1080,Cairo,Data Science,Male,25,14.1,73.2,7.7,0,4,Yes,Bus,0,NaN,79.3
3,ITI-1128,Cairo,Web Development,Female,25,-5,13.7,6.3,2,4,Yes,Metro,1,NaN,90.6
4,ITI-1197,Menoufia,Web Development,Male,25,19.9,17.2,6.9,3,3,No,Car,0,none,86.8


In [6]:
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271 entries, 0 to 270
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   student_id          271 non-null    object 
 1   branch              271 non-null    object 
 2   track               271 non-null    object 
 3   gender              271 non-null    object 
 4   age                 269 non-null    object 
 5   study_hours_week    264 non-null    object 
 6   commute_minutes     265 non-null    object 
 7   sleep_hours         260 non-null    float64
 8   projects_submitted  271 non-null    int64  
 9   satisfaction_1to5   266 non-null    object 
 10  has_laptop          271 non-null    object 
 11  transport           271 non-null    object 
 12  pets_at_home        271 non-null    int64  
 13  favorite_sport      243 non-null    object 
 14  final_score         266 non-null    object 
dtypes: float64(1), int64(2), object(12)
memory usage: 31.9+ K

In [7]:
df_raw["student_id"].value_counts()

,count
student_id,
ITI-1103,2
ITI-1068,2
ITI-1181,2
ITI-1109,2
ITI-1032,2
...,...
ITI-1070,1
ITI-1003,1
ITI-1077,1


In [8]:
df_raw["student_id"].nunique()

261

In [9]:
df_raw["student_id"].isna().sum()

np.int64(0)

In [10]:
df_raw["student_id"].duplicated().sum()

np.int64(10)

In [11]:
df_raw["student_id"].sort_values().to_list()

[' ITI-1005 ',
 ' ITI-1007 ',
 ' ITI-1046 ',
 ' ITI-1062 ',
 ' ITI-1093 ',
 ' ITI-1119 ',
 ' ITI-1155 ',
 ' ITI-1232 ',
 ' ITI-1236 ',
 ' ITI-1259 ',
 'ITI-1001',
 'ITI-1002',
 'ITI-1003',
 'ITI-1004',
 'ITI-1006',
 'ITI-1008',
 'ITI-1009',
 'ITI-1010',
 'ITI-1011',
 'ITI-1012',
 'ITI-1013',
 'ITI-1014',
 'ITI-1015',
 'ITI-1016',
 'ITI-1017',
 'ITI-1017',
 'ITI-1018',
 'ITI-1019',
 'ITI-1020',
 'ITI-1021',
 'ITI-1022',
 'ITI-1023',
 'ITI-1024',
 'ITI-1025',
 'ITI-1026',
 'ITI-1027',
 'ITI-1028',
 'ITI-1029',
 'ITI-1030',
 'ITI-1031',
 'ITI-1032',
 'ITI-1032',
 'ITI-1033',
 'ITI-1034',
 'ITI-1035',
 'ITI-1036',
 'ITI-1037',
 'ITI-1038',
 'ITI-1039',
 'ITI-1040',
 'ITI-1041',
 'ITI-1042',
 'ITI-1042',
 'ITI-1043',
 'ITI-1044',
 'ITI-1045',
 'ITI-1047',
 'ITI-1048',
 'ITI-1049',
 'ITI-1050',
 'ITI-1051',
 'ITI-1052',
 'ITI-1053',
 'ITI-1054',
 'ITI-1055',
 'ITI-1056',
 'ITI-1057',
 'ITI-1058',
 'ITI-1059',
 'ITI-1060',
 'ITI-1061',
 'ITI-1062',
 'ITI-1063',
 'ITI-1064',
 'ITI-1065',
 'ITI

In [12]:
duplicate_ids=df_raw["student_id"].value_counts()[df_raw["student_id"].value_counts()>1].index
duplicate_ids

Index(['ITI-1103', 'ITI-1068', 'ITI-1181', 'ITI-1109', 'ITI-1032', 'ITI-1042',
       'ITI-1096', 'ITI-1244', 'ITI-1214', 'ITI-1017'],
      dtype='object', name='student_id')

In [13]:
df_duplicates =df_raw[df_raw["student_id"].isin(duplicate_ids)].sort_values("student_id")
df_duplicates

,student_id,branch,track,gender,age,study_hours_week,commute_minutes,sleep_hours,projects_submitted,satisfaction_1to5,has_laptop,transport,pets_at_home,favorite_sport,final_score
117,ITI-1017,Cairo,Cyber Security,Male,23,21.5,59.6,8.5,8,3,Yes,Walk,0,NaN,85.9
60,ITI-1017,Cairo,Cyber Security,Male,23,21.5,59.6,8.5,8,3,Yes,Walk,0,none,85.9
126,ITI-1032,Cairo,Embedded Systems,M,23,20.5,62.9,5.1,3,5,Yes,Microbus,2,Volleyball,84.6
92,ITI-1032,Cairo,Embedded Systems,M,23,20.5,62.9,5.1,3,5,Yes,Microbus,2,Volleyball,84.6
135,ITI-1042,cairo,Data Science,Male,27,21.9,22.8,6.7,5,5,Yes,Bus,0,Basketball,87.7
11,ITI-1042,cairo,Data Science,Male,27,21.9,22.8,6.7,6,5,Yes,Bus,0,Basketball,87.7
106,ITI-1068,Menoufia,Data Science,Male,25,16.6,44.1,6.6,3,5,No,Bus,0,Football,79.5
37,ITI-1068,Menoufia,Data Science,Male,25,16.6,44.1,6.6,3,5,No,Bus,0,Football,79.5
132,ITI-1096,Menoufia,Data Science,Male,27,24.8,29.3,7.0,4,4,Yes,Microbus,0,-,93.0
155,ITI-1096,Menoufia,Data Science,Male,27,24.8,29.3,7.0,5,4,Yes,Microbus,0,-,93.0


In [14]:
valid_ids= {f"ITI-{i}" for i in range(1001,1261)}

In [15]:
invalid_ids=sorted(set(df_raw['student_id']) - valid_ids)
invalid_ids

[' ITI-1005 ',
 ' ITI-1007 ',
 ' ITI-1046 ',
 ' ITI-1062 ',
 ' ITI-1093 ',
 ' ITI-1119 ',
 ' ITI-1155 ',
 ' ITI-1232 ',
 ' ITI-1236 ',
 ' ITI-1259 ']

### White spaces before and after the student Id

In [16]:
all_ids = sorted(df_raw["student_id"].dropna().unique())

print(all_ids)

[' ITI-1005 ', ' ITI-1007 ', ' ITI-1046 ', ' ITI-1062 ', ' ITI-1093 ', ' ITI-1119 ', ' ITI-1155 ', ' ITI-1232 ', ' ITI-1236 ', ' ITI-1259 ', 'ITI-1001', 'ITI-1002', 'ITI-1003', 'ITI-1004', 'ITI-1006', 'ITI-1008', 'ITI-1009', 'ITI-1010', 'ITI-1011', 'ITI-1012', 'ITI-1013', 'ITI-1014', 'ITI-1015', 'ITI-1016', 'ITI-1017', 'ITI-1018', 'ITI-1019', 'ITI-1020', 'ITI-1021', 'ITI-1022', 'ITI-1023', 'ITI-1024', 'ITI-1025', 'ITI-1026', 'ITI-1027', 'ITI-1028', 'ITI-1029', 'ITI-1030', 'ITI-1031', 'ITI-1032', 'ITI-1033', 'ITI-1034', 'ITI-1035', 'ITI-1036', 'ITI-1037', 'ITI-1038', 'ITI-1039', 'ITI-1040', 'ITI-1041', 'ITI-1042', 'ITI-1043', 'ITI-1044', 'ITI-1045', 'ITI-1047', 'ITI-1048', 'ITI-1049', 'ITI-1050', 'ITI-1051', 'ITI-1052', 'ITI-1053', 'ITI-1054', 'ITI-1055', 'ITI-1056', 'ITI-1057', 'ITI-1058', 'ITI-1059', 'ITI-1060', 'ITI-1061', 'ITI-1062', 'ITI-1063', 'ITI-1064', 'ITI-1065', 'ITI-1066', 'ITI-1067', 'ITI-1068', 'ITI-1069', 'ITI-1070', 'ITI-1071', 'ITI-1072', 'ITI-1073', 'ITI-1074', 'ITI-10

In [17]:
df_raw["student_id"].str.strip().unique()

array(['ITI-1031', 'ITI-1117', 'ITI-1080', 'ITI-1128', 'ITI-1197',
       'ITI-1138', 'ITI-1210', 'ITI-1046', 'ITI-1159', 'ITI-1248',
       'ITI-1184', 'ITI-1042', 'ITI-1228', 'ITI-1083', 'ITI-1166',
       'ITI-1195', 'ITI-1227', 'ITI-1147', 'ITI-1105', 'ITI-1061',
       'ITI-1222', 'ITI-1109', 'ITI-1047', 'ITI-1043', 'ITI-1186',
       'ITI-1010', 'ITI-1023', 'ITI-1200', 'ITI-1110', 'ITI-1025',
       'ITI-1114', 'ITI-1069', 'ITI-1145', 'ITI-1225', 'ITI-1253',
       'ITI-1007', 'ITI-1121', 'ITI-1068', 'ITI-1120', 'ITI-1119',
       'ITI-1026', 'ITI-1126', 'ITI-1245', 'ITI-1020', 'ITI-1078',
       'ITI-1217', 'ITI-1091', 'ITI-1209', 'ITI-1094', 'ITI-1181',
       'ITI-1016', 'ITI-1153', 'ITI-1233', 'ITI-1251', 'ITI-1116',
       'ITI-1034', 'ITI-1185', 'ITI-1143', 'ITI-1198', 'ITI-1038',
       'ITI-1017', 'ITI-1076', 'ITI-1149', 'ITI-1011', 'ITI-1234',
       'ITI-1067', 'ITI-1254', 'ITI-1226', 'ITI-1183', 'ITI-1074',
       'ITI-1019', 'ITI-1102', 'ITI-1196', 'ITI-1113', 'ITI-11

In [18]:
temp_ids = df_raw["student_id"].str.strip()

duplicate_ids = temp_ids.value_counts()
duplicate_ids[duplicate_ids > 1]

,count
student_id,
ITI-1062,2
ITI-1103,2
ITI-1068,2
ITI-1181,2
ITI-1109,2
ITI-1032,2
ITI-1042,2
ITI-1244,2
ITI-1096,2


In [19]:
duplicate_ids = duplicate_ids[duplicate_ids > 1].index

df_duplicates = df_raw[temp_ids.isin(duplicate_ids)].copy()

df_duplicates["student_id_clean"] = df_duplicates["student_id"].str.strip()

df_duplicates = df_duplicates.sort_values("student_id_clean")

df_duplicates

,student_id,branch,track,gender,age,study_hours_week,commute_minutes,sleep_hours,projects_submitted,satisfaction_1to5,has_laptop,transport,pets_at_home,favorite_sport,final_score,student_id_clean
117,ITI-1017,Cairo,Cyber Security,Male,23,21.5,59.6,8.5,8,3,Yes,Walk,0,NaN,85.9,ITI-1017
60,ITI-1017,Cairo,Cyber Security,Male,23,21.5,59.6,8.5,8,3,Yes,Walk,0,none,85.9,ITI-1017
126,ITI-1032,Cairo,Embedded Systems,M,23,20.5,62.9,5.1,3,5,Yes,Microbus,2,Volleyball,84.6,ITI-1032
92,ITI-1032,Cairo,Embedded Systems,M,23,20.5,62.9,5.1,3,5,Yes,Microbus,2,Volleyball,84.6,ITI-1032
135,ITI-1042,cairo,Data Science,Male,27,21.9,22.8,6.7,5,5,Yes,Bus,0,Basketball,87.7,ITI-1042
11,ITI-1042,cairo,Data Science,Male,27,21.9,22.8,6.7,6,5,Yes,Bus,0,Basketball,87.7,ITI-1042
114,ITI-1062,Mansoura,Embedded Systems,Female,23,14.0,23.3,7.6,4,5,Yes,Walk,0,Basketball,76.7,ITI-1062
150,ITI-1062,Mansoura,Embedded Systems,Female,23,14.0,23.3,7.6,4,5,Yes,Walk,0,Basketball,76.7,ITI-1062
106,ITI-1068,Menoufia,Data Science,Male,25,16.6,44.1,6.6,3,5,No,Bus,0,Football,79.5,ITI-1068
37,ITI-1068,Menoufia,Data Science,Male,25,16.6,44.1,6.6,3,5,No,Bus,0,Football,79.5,ITI-1068


In [20]:
df_raw[df_raw["student_id"].str.strip().isin([
    "ITI-1017",
    "ITI-1042",
    "ITI-1096",
    "ITI-1103",
    "ITI-1109",
    "ITI-1181"
])]

,student_id,branch,track,gender,age,study_hours_week,commute_minutes,sleep_hours,projects_submitted,satisfaction_1to5,has_laptop,transport,pets_at_home,favorite_sport,final_score
11,ITI-1042,cairo,Data Science,Male,27,21.9,22.8,6.7,6,5,Yes,Bus,0,Basketball,87.7
21,ITI-1109,Menoufia,Data Science,Male,23,9.5,43.9,6.6,3,5,Yes,Microbus,1,Basketball,77.0
49,ITI-1181,Assiut,Data Science,Male,22,11.3,11.8,7.1,6,4,Yes,Microbus,1,Football,82.1
60,ITI-1017,Cairo,Cyber Security,Male,23,21.5,59.6,8.5,8,3,Yes,Walk,0,none,85.9
103,ITI-1109,Menoufia,Data Science,Male,23,9.5,43.9,6.6,2,5,Yes,Microbus,1,Basketball,77.0
117,ITI-1017,Cairo,Cyber Security,Male,23,21.5,59.6,8.5,8,3,Yes,Walk,0,NaN,85.9
132,ITI-1096,Menoufia,Data Science,Male,27,24.8,29.3,7.0,4,4,Yes,Microbus,0,-,93.0
135,ITI-1042,cairo,Data Science,Male,27,21.9,22.8,6.7,5,5,Yes,Bus,0,Basketball,87.7
155,ITI-1096,Menoufia,Data Science,Male,27,24.8,29.3,7.0,5,4,Yes,Microbus,0,-,93.0
254,ITI-1181,Assiut,Data Science,Male,22,11.3,11.8,7.1,7,4,Yes,Microbus,1,Football,82.1


In [21]:
df_check = df_raw.copy()

df_check["student_id"] = df_check["student_id"].str.strip()

df_check.duplicated(subset="student_id", keep=False).sum()

np.int64(22)

In [22]:
df_check = df_raw.copy()
df_check["student_id"] = df_check["student_id"].str.strip()

df_check.duplicated(keep=False).sum()

np.int64(10)

In [23]:
exact_duplicates = df_check[df_check.duplicated(keep=False)].sort_values("student_id")

exact_duplicates

,student_id,branch,track,gender,age,study_hours_week,commute_minutes,sleep_hours,projects_submitted,satisfaction_1to5,has_laptop,transport,pets_at_home,favorite_sport,final_score
92,ITI-1032,Cairo,Embedded Systems,M,23,20.5,62.9,5.1,3,5,Yes,Microbus,2,Volleyball,84.6
126,ITI-1032,Cairo,Embedded Systems,M,23,20.5,62.9,5.1,3,5,Yes,Microbus,2,Volleyball,84.6
114,ITI-1062,Mansoura,Embedded Systems,Female,23,14.0,23.3,7.6,4,5,Yes,Walk,0,Basketball,76.7
150,ITI-1062,Mansoura,Embedded Systems,Female,23,14.0,23.3,7.6,4,5,Yes,Walk,0,Basketball,76.7
37,ITI-1068,Menoufia,Data Science,Male,25,16.6,44.1,6.6,3,5,No,Bus,0,Football,79.5
106,ITI-1068,Menoufia,Data Science,Male,25,16.6,44.1,6.6,3,5,No,Bus,0,Football,79.5
140,ITI-1214,Cairo,Web Development,Male,22,16.2,13.4,8.1,2,NaN,Yes,Metro,0,Football,98.8
183,ITI-1214,Cairo,Web Development,Male,22,16.2,13.4,8.1,2,NaN,Yes,Metro,0,Football,98.8
170,ITI-1244,Cairo,Web Development,Female,20,22.6,97.9,6.7,1,4,Yes,Microbus,0,Football,80.7
171,ITI-1244,Cairo,Web Development,Female,20,22.6,97.9,6.7,1,4,Yes,Microbus,0,Football,80.7


### Make a copy for cleand Dataset

In [24]:
df_clean = df_raw.copy()

### Clean the spacing before and after the student Id

In [25]:
df_clean["student_id"] = df_clean["student_id"].str.strip()

In [26]:
df_clean["student_id"].nunique()

260

### Drop (Dlete) the exact rows(IDs)

In [27]:
df_clean = df_clean.drop_duplicates()

In [28]:
df_clean.shape

(266, 15)

In [29]:
df_clean[df_clean["student_id"] == "ITI-1042"]

,student_id,branch,track,gender,age,study_hours_week,commute_minutes,sleep_hours,projects_submitted,satisfaction_1to5,has_laptop,transport,pets_at_home,favorite_sport,final_score
11,ITI-1042,cairo,Data Science,Male,27,21.9,22.8,6.7,6,5,Yes,Bus,0,Basketball,87.7
135,ITI-1042,cairo,Data Science,Male,27,21.9,22.8,6.7,5,5,Yes,Bus,0,Basketball,87.7


In [30]:
near_dup_ids = [
    "ITI-1017",
    "ITI-1042",
    "ITI-1096",
    "ITI-1103",
    "ITI-1109",
    "ITI-1181"
]

df_clean[df_clean["student_id"].isin(near_dup_ids)].sort_values("student_id")

,student_id,branch,track,gender,age,study_hours_week,commute_minutes,sleep_hours,projects_submitted,satisfaction_1to5,has_laptop,transport,pets_at_home,favorite_sport,final_score
60,ITI-1017,Cairo,Cyber Security,Male,23,21.5,59.6,8.5,8,3,Yes,Walk,0,none,85.9
117,ITI-1017,Cairo,Cyber Security,Male,23,21.5,59.6,8.5,8,3,Yes,Walk,0,NaN,85.9
11,ITI-1042,cairo,Data Science,Male,27,21.9,22.8,6.7,6,5,Yes,Bus,0,Basketball,87.7
135,ITI-1042,cairo,Data Science,Male,27,21.9,22.8,6.7,5,5,Yes,Bus,0,Basketball,87.7
132,ITI-1096,Menoufia,Data Science,Male,27,24.8,29.3,7.0,4,4,Yes,Microbus,0,-,93.0
155,ITI-1096,Menoufia,Data Science,Male,27,24.8,29.3,7.0,5,4,Yes,Microbus,0,-,93.0
265,ITI-1103,Alexandria,Embedded Systems,Male,21,17.4,64.1,8.5,4,4,Yes,Car,0,Basketball,81.9
270,ITI-1103,Alexandria,Embedded Systems,Male,21,17.4,64.1,8.5,3,4,Yes,Car,0,Basketball,81.9
21,ITI-1109,Menoufia,Data Science,Male,23,9.5,43.9,6.6,3,5,Yes,Microbus,1,Basketball,77.0
103,ITI-1109,Menoufia,Data Science,Male,23,9.5,43.9,6.6,2,5,Yes,Microbus,1,Basketball,77.0


In [31]:
df_clean[df_clean["student_id"] == "ITI-1017"]

,student_id,branch,track,gender,age,study_hours_week,commute_minutes,sleep_hours,projects_submitted,satisfaction_1to5,has_laptop,transport,pets_at_home,favorite_sport,final_score
60,ITI-1017,Cairo,Cyber Security,Male,23,21.5,59.6,8.5,8,3,Yes,Walk,0,none,85.9
117,ITI-1017,Cairo,Cyber Security,Male,23,21.5,59.6,8.5,8,3,Yes,Walk,0,NaN,85.9


### Drop the student id "ITI-1017" the second row as is the same but there is a difference in the favorite_sport

In [32]:
mask=df_clean["student_id"]=="ITI-1017"
first_idx= df_clean[mask].index[0]
df_clean=df_clean.drop(index=df_clean[mask].index[1:])

In [33]:
df_clean[df_clean["student_id"] == "ITI-1017"]

,student_id,branch,track,gender,age,study_hours_week,commute_minutes,sleep_hours,projects_submitted,satisfaction_1to5,has_laptop,transport,pets_at_home,favorite_sport,final_score
60,ITI-1017,Cairo,Cyber Security,Male,23,21.5,59.6,8.5,8,3,Yes,Walk,0,none,85.9


In [34]:
df_clean.shape

(265, 15)

### Inconsistent categories

In [35]:
df_clean["branch"].value_counts(dropna=False)

,count
branch,
Cairo,95
Alexandria,51
Mansoura,42
Menoufia,34
Assiut,32
cairo,4
alexandria,4
CAIRO,2
Mansora,1


In [36]:
df_clean["branch"].unique()

array(['Cairo', 'Assiut', 'Menoufia', 'Alexandria', 'Mansoura', 'cairo',
       'alexandria', 'Mansora', 'CAIRO'], dtype=object)

In [37]:
df_clean["branch"] = df_clean["branch"].replace({
    "cairo": "Cairo",
    "CAIRO": "Cairo",
    "alexandria": "Alexandria",
    "Mansora": "Mansoura"
})

In [38]:
df_clean["branch"].value_counts()

,count
branch,
Cairo,101
Alexandria,55
Mansoura,43
Menoufia,34
Assiut,32


In [39]:
df_clean["gender"].value_counts(dropna=False)

,count
gender,
Male,145
Female,108
M,5
female,3
F,2
MALE,2


In [40]:
df_clean["gender"] = df_clean["gender"].replace({
    "M": "Male",
    "MALE": "Male",
    "female": "Female",
    "F": "Female"
})

In [41]:
df_clean["gender"].value_counts(dropna=False)

,count
gender,
Male,152
Female,113


In [42]:
df_clean["track"].value_counts(dropna=False)

,count
track,
Data Science,85
Web Development,64
Cyber Security,38
UI/UX,33
Embedded Systems,31
web development,3
data science,2
cyber security,2
Embedded Systems,2


#### There is a space in the " UI/UX ", "Data  Science", " Web Development "," Embedded Systems "

In [43]:
df_clean["track"].unique()

array(['Data Science', 'Web Development', 'UI/UX', 'web development',
       'data science', 'Embedded Systems', 'Cyber Security',
       'cyber security', ' UI/UX ', 'WEB DEVELOPMENT', 'Data  Science',
       ' Web Development ', ' Embedded Systems '], dtype=object)

In [44]:
df_clean["track"].apply(lambda x: x != x.strip() if isinstance(x, str) else False).sum()

np.int64(5)

In [45]:
valid_tracks = [
    "Data Science",
    "Web Development",
    "Cyber Security",
    "UI/UX",
    "Embedded Systems"
]

df_raw.loc[
    ~df_raw["track"].isin(valid_tracks),
    "track"
].value_counts(dropna=False)

,count
track,
web development,3
data science,2
cyber security,2
Embedded Systems,2
Web Development,2
UI/UX,1
Data Science,1
WEB DEVELOPMENT,1


In [46]:
df_clean["track"] = df_clean["track"].str.strip()

In [47]:
df_clean["track"] = df_clean["track"].replace({
    "web development": "Web Development",
    "WEB DEVELOPMENT": "Web Development",
    "data science": "Data Science",
    "cyber security": "Cyber Security",
    "Data  Science": "Data Science"
})

In [48]:
df_clean["track"].value_counts(dropna=False)

,count
track,
Data Science,88
Web Development,70
Cyber Security,40
UI/UX,34
Embedded Systems,33


In [49]:
df_clean["transport"].value_counts(dropna=False)

,count
transport,
Bus,86
Metro,59
Car,51
Microbus,37
Walk,24
bus,4
METRO,2
micro bus,1
car,1


In [50]:
df_clean[df_clean["transport"].str.lower().str.contains("car", na=False)]["transport"].unique()

array(['Car', 'car '], dtype=object)

In [51]:
df_clean["transport"] = df_clean["transport"].str.strip().str.title()

In [52]:
df_clean["transport"] = df_clean["transport"].replace({
    "bus": "Bus",
    "METRO": "Metro",
    "micro bus": "Microbus",
    "car": "Car"
})

In [53]:
df_clean["transport"].value_counts(dropna=False)

,count
transport,
Bus,90
Metro,61
Car,52
Microbus,37
Walk,24
Micro Bus,1


In [54]:
df_clean["transport"] = (
    df_raw["transport"].astype(str).str.strip().str.title()
    .replace({"Micro Bus": "Microbus"})
)

In [55]:
df_clean["favorite_sport"].value_counts(dropna=False)

,count
favorite_sport,
Football,85
Basketball,39
Running,38
Swimming,33
NaN,27
No sport,15
Volleyball,14
none,11
-,3


In [56]:
df_clean["favorite_sport"] = df_clean["favorite_sport"].replace({
    "No sport": "None",
    "none": "None"
})

In [57]:
df_clean["favorite_sport"].value_counts(dropna=False)

,count
favorite_sport,
Football,85
Basketball,39
Running,38
Swimming,33
NaN,27
None,26
Volleyball,14
-,3


In [58]:
df_raw[df_raw["favorite_sport"] == "-"]

,student_id,branch,track,gender,age,study_hours_week,commute_minutes,sleep_hours,projects_submitted,satisfaction_1to5,has_laptop,transport,pets_at_home,favorite_sport,final_score
132,ITI-1096,Menoufia,Data Science,Male,27,24.8,29.3,7.0,4,4,Yes,Microbus,0,-,93.0
155,ITI-1096,Menoufia,Data Science,Male,27,24.8,29.3,7.0,5,4,Yes,Microbus,0,-,93.0
233,ITI-1055,Menoufia,UI/UX,Female,24,8.0,16.8,6.8,2,4,Yes,Walk,3,-,73.3


In [59]:
df_clean.loc[
    df_clean["student_id"] == "ITI-1096",
    "favorite_sport"
] = "None"

In [60]:
df_clean.loc[
    df_clean["student_id"] == "ITI-1096",
    ["student_id", "favorite_sport"]
]

,student_id,favorite_sport
132,ITI-1096,None
155,ITI-1096,None


In [61]:
df_clean["favorite_sport"] = df_clean["favorite_sport"].replace("-", np.nan)

In [62]:
df_clean["favorite_sport"].value_counts(dropna=False)

,count
favorite_sport,
Football,85
Basketball,39
Running,38
Swimming,33
NaN,28
None,28
Volleyball,14


In [63]:
df_raw["has_laptop"].value_counts(dropna=False)

,count
has_laptop,
Yes,211
No,51
1,3
Y,2
NO,2
N,1
yes,1


In [64]:
df_clean["has_laptop"] = df_clean["has_laptop"].replace({
    "1": "Yes",
    "Y": "Yes",
    "NO": "No",
    "N": "No",
    "yes": "Yes"
})

In [65]:
df_clean["has_laptop"].value_counts(dropna=False)

,count
has_laptop,
Yes,212
No,53


In [66]:
df_clean["pets_at_home"].value_counts(dropna=False)

,count
pets_at_home,
0,154
1,56
2,33
3,15
4,4
-2,2
60,1


In [67]:
df_raw[
    (df_raw["pets_at_home"] < 0) |
    (df_raw["pets_at_home"] > 20)
]

,student_id,branch,track,gender,age,study_hours_week,commute_minutes,sleep_hours,projects_submitted,satisfaction_1to5,has_laptop,transport,pets_at_home,favorite_sport,final_score
42,ITI-1245,Cairo,Data Science,Male,24,18.8,26.4,NaN,1,3,Yes,Bus,60,Football,75.2
44,ITI-1078,Alexandria,Web Development,Male,28,11.0,78.0,8.5,1,1,Yes,Walk,-2,Basketball,74.5
134,ITI-1180,Menoufia,WEB DEVELOPMENT,Female,28,18.8,29.5,6.4,0,4,Yes,Metro,-2,Basketball,75.7


In [68]:
df_clean.loc[
    (df_clean["pets_at_home"] < 0) |
    (df_clean["pets_at_home"] > 20),
    "pets_at_home"
] = np.nan

In [69]:
df_clean[
    (df_clean["pets_at_home"] < 0) |
    (df_clean["pets_at_home"] > 20)
]

,student_id,branch,track,gender,age,study_hours_week,commute_minutes,sleep_hours,projects_submitted,satisfaction_1to5,has_laptop,transport,pets_at_home,favorite_sport,final_score


In [70]:
df_raw["satisfaction_1to5"].value_counts(dropna=False)

,count
satisfaction_1to5,
4,92
3,72
5,55
2,41
NaN,5
?,2
7,2
1,1
6,1


In [71]:
df_raw[
    (~df_raw["satisfaction_1to5"].isin([1, 2, 3, 4, 5])) &
    (df_raw["satisfaction_1to5"].notna())
]

,student_id,branch,track,gender,age,study_hours_week,commute_minutes,sleep_hours,projects_submitted,satisfaction_1to5,has_laptop,transport,pets_at_home,favorite_sport,final_score
0,ITI-1031,Cairo,Data Science,Male,21,16.5,34.1,5.5,2,3,Yes,Car,0,Swimming,79.3
1,ITI-1117,Assiut,Data Science,Female,29,25.1 hours,5.0,6.6,3,4,Yes,Metro,1,Football,91.7
2,ITI-1080,Cairo,Data Science,Male,25,14.1,73.2,7.7,0,4,Yes,Bus,0,NaN,79.3
3,ITI-1128,Cairo,Web Development,Female,25,-5,13.7,6.3,2,4,Yes,Metro,1,NaN,90.6
4,ITI-1197,Menoufia,Web Development,Male,25,19.9,17.2,6.9,3,3,No,Car,0,none,86.8
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
266,ITI-1021,Cairo,web development,Male,23,17.1,46.4,8.9,1,4,No,Bus,3,Football,89.4
267,ITI-1189,Mansoura,Data Science,Male,26,12.9,NaN,7.0,2,2,NO,Microbus,1,Basketball,76.5
268,ITI-1072,Mansoura,Embedded Systems,Male,25,24.7,34.1,8.5,4,5,Yes,METRO,3,Running,96.4
269,ITI-1107,Assiut,Data Science,Female,24,-5,29.8,8.4,2,3,Yes,Bus,0,Football,85.4


In [72]:
(
    (~df_raw["satisfaction_1to5"].isin([1, 2, 3, 4, 5])) &
    (df_raw["satisfaction_1to5"].notna())
).sum()

np.int64(266)

In [73]:
df_raw["satisfaction_1to5"].apply(type).value_counts()

,count
satisfaction_1to5,
<class 'str'>,266
<class 'float'>,5


In [74]:
df_clean["satisfaction_1to5"].unique()[:10]

df_clean["satisfaction_1to5"] = pd.to_numeric(
    df_clean["satisfaction_1to5"], errors="coerce"
)

In [75]:
df_clean.loc[
    ~df_clean["satisfaction_1to5"].isin([1, 2, 3, 4, 5]),
    "satisfaction_1to5"
] = np.nan


In [76]:
df_clean["satisfaction_1to5"].value_counts(dropna=False)

,count
satisfaction_1to5,
4.0,91
3.0,71
5.0,52
2.0,41
NaN,9
1.0,1


In [77]:
df_raw["age"].describe()

,age
count,269
unique,15
top,23
freq,54


In [78]:
age_numeric = pd.to_numeric(df_raw["age"], errors="coerce")

In [79]:
invalid_age = (
    (age_numeric < 17) |
    (age_numeric > 45)
)

invalid_age.sum()

np.int64(4)

In [80]:
df_raw.loc[invalid_age, ["student_id", "age"]]

,student_id,age
32,ITI-1145,150
97,ITI-1003,-1
204,ITI-1164,5
219,ITI-1009,-1


In [81]:
df_clean.loc[
    (pd.to_numeric(df_clean["age"], errors="coerce") < 17) |
    (pd.to_numeric(df_clean["age"], errors="coerce") > 45) |
    (pd.to_numeric(df_clean["age"], errors="coerce")== "unknown"),
    "age"
] = np.nan

In [82]:
df_clean["age"].value_counts(dropna=False).sort_index()

,count
age,
20,23
21,32
22,28
23,51
24,41
25,38
26,16
27,10
28,13


In [83]:
df_raw["study_hours_week"].loc[[1, 14, 116, 137, 205, 225]]

,study_hours_week
1,25.1 hours
14,13.1 hours
116,23.3 hours
137,6.8 hours
205,11.0 hours
225,21.6 hours


In [84]:
df_clean["study_hours_week"] = (
    df_raw["study_hours_week"]
    .astype(str)
    .str.replace(" hours", "", regex=False)
    .replace("missing", np.nan)
)

df_clean["study_hours_week"] = pd.to_numeric(
    df_clean["study_hours_week"],
    errors="coerce"
)

In [85]:
df_clean.loc[
    (df_clean["study_hours_week"] < 0) |
    (df_clean["study_hours_week"] > 80),
    "study_hours_week"
] = np.nan

In [86]:
df_clean.loc[
    df_raw["study_hours_week"].astype(str).str.contains("hours", na=False),
    ["student_id", "study_hours_week"]
]

,student_id,study_hours_week
1,ITI-1117,25.1
14,ITI-1166,13.1
116,ITI-1057,23.3
137,ITI-1177,6.8
205,ITI-1081,11.0
225,ITI-1167,21.6


In [87]:
commute_numeric = pd.to_numeric(
    df_raw["commute_minutes"],
    errors="coerce"
)

invalid_commute = (
    (commute_numeric < 0) |
    (commute_numeric > 300)
)

invalid_commute.sum()

np.int64(0)

In [88]:
df_raw["commute_minutes"].apply(type).value_counts()

,count
commute_minutes,
<class 'str'>,265
<class 'float'>,6


In [89]:
df_raw["commute_minutes"].isna().sum()

np.int64(6)

In [90]:
commute_numeric = pd.to_numeric(
    df_raw["commute_minutes"],
    errors="coerce"
)

df_raw.loc[
    commute_numeric.isna() & df_raw["commute_minutes"].notna(),
    ["student_id", "commute_minutes"]
]

,student_id,commute_minutes
28,ITI-1110,31 min
31,ITI-1069,83 min
89,ITI-1112,50 min
102,ITI-1097,67 min
129,ITI-1029,43 min
178,ITI-1041,35 min
210,ITI-1172,46 min
262,ITI-1088,70 min


In [91]:
mapping= {
    "31 min":31.0,
    "83 min":83.0,
    "50 min":50.0,
    "67 min":67.0,
    "43 min":43.0,
    "35 min":35.0,
    "46 min":46.0,
    "70 min":70.0
}

df_clean.replace(mapping, inplace=True)

In [92]:
df_clean["commute_minutes"].apply(type).value_counts()

,count
commute_minutes,
<class 'str'>,251
<class 'float'>,14


In [93]:
commute_numeric = pd.to_numeric(
    df_clean["commute_minutes"],
    errors="coerce"
)

df_clean.loc[
    commute_numeric.isna() & df_clean["commute_minutes"].notna(),
    ["student_id", "commute_minutes"]
]

,student_id,commute_minutes


In [94]:
df_clean["final_score"] = pd.to_numeric(
    df_raw["final_score"].astype(str).str.replace(",", ".", regex=False),
    errors="coerce"
)
df_clean.loc[
    (df_clean["final_score"] < 0) | (df_clean["final_score"] > 100),
    "final_score"
] = np.nan

df_clean["sleep_hours"] = pd.to_numeric(
    df_clean["sleep_hours"], errors="coerce"
)
df_clean.loc[
    (df_clean["sleep_hours"] < 2) | (df_clean["sleep_hours"] > 14),
    "sleep_hours"
] = np.nan


df_clean["projects_submitted"] = pd.to_numeric(
    df_clean["projects_submitted"], errors="coerce"
)

for col, (lo, hi) in {
    "sleep_hours":        (2, 14),
    "projects_submitted": (0, 12),
    "final_score":        (0, 100),
}.items():
    s = df_clean[col]
    print(f"{col:20s}  min={s.min():>6}  max={s.max():>6}  "
          f"NaN={s.isna().sum():>3}  out-of-range={((s < lo) | (s > hi)).sum()}")

sleep_hours           min=   3.8  max=   9.5  NaN= 15  out-of-range=0
projects_submitted    min=     0  max=     8  NaN=  0  out-of-range=0
final_score           min=  50.4  max= 100.0  NaN=  8  out-of-range=0


In [95]:
df_clean.isnull().sum()

,0
student_id,0
branch,0
track,0
gender,0
age,6
study_hours_week,12
commute_minutes,6
sleep_hours,15
projects_submitted,0
satisfaction_1to5,9


### converting DataTypes
#### Identifier

In [96]:
df_clean["student_id"] = df_clean["student_id"].astype("string")

#### Qualitative — Nominal

In [97]:
nominal_cols = {
    "branch":         ["Cairo", "Alexandria", "Mansoura", "Assiut", "Menoufia"],
    "track":          ["Data Science", "Web Development", "Cyber Security",
                       "Embedded Systems", "UI/UX"],
    "gender":         ["Male", "Female"],
    "has_laptop":     ["Yes", "No"],
    "transport":      ["Bus", "Metro", "Car", "Walk", "Microbus"],
    "favorite_sport": ["Football", "Swimming", "Basketball",
                       "Running", "Volleyball", "None"],
}
for col, order in nominal_cols.items():
    df_clean[col] = pd.Categorical(df_clean[col], categories=order, ordered=False)


#### Qualitative — Ordinal

In [98]:
df_clean["satisfaction_1to5"] = pd.Categorical(
    df_clean["satisfaction_1to5"],
    categories=[1, 2, 3, 4, 5],
    ordered=True,
)

df_clean["satisfaction_1to5"].value_counts(dropna=False).sort_index()

,count
satisfaction_1to5,
1,1
2,41
3,71
4,91
5,52
NaN,9


#### Quantitative — Discrete

In [99]:
for col in ["age", "projects_submitted", "pets_at_home"]:
    df_clean[col] = pd.to_numeric(df_clean[col], errors="coerce")
    df_clean[col] = df_clean[col].astype("Int64")

#### Quantitative — Continuous (float)

In [100]:
for col in ["study_hours_week", "commute_minutes", "sleep_hours", "final_score"]:
    df_clean[col] = df_clean[col].astype("float64")

## Final clean dataset (df_clean)
### Shape: 265 × 15
### All dtypes match data_dictionary
### All NaN values documented in cleaning_log

In [101]:
print("Shape:", df_clean.shape)
print("\nNulls:\n", df_clean.isna().sum())
print("\ndtypes:\n", df_clean.dtypes)

Shape: (265, 15)

Nulls:
 student_id             0
branch                 0
track                  0
gender                 0
age                    8
study_hours_week      12
commute_minutes        6
sleep_hours           15
projects_submitted     0
satisfaction_1to5      9
has_laptop             0
transport              0
pets_at_home           3
favorite_sport        28
final_score            8
dtype: int64

dtypes:
 student_id            string[python]
branch                      category
track                       category
gender                      category
age                            Int64
study_hours_week             float64
commute_minutes              float64
sleep_hours                  float64
projects_submitted             Int64
satisfaction_1to5           category
has_laptop                  category
transport                   category
pets_at_home                   Int64
favorite_sport              category
final_score                  float64
dtype: object


In [103]:
df_clean.to_excel("cleaned_data.xlsx", index=False)

In [104]:
from google.colab import files
files.download("cleaned_data.xlsx")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>